File: Data Structures & Algorithms/max-water-container/submission-0.java
class Solution {
    public int maxArea(int[] heights) {
        int left = 0;
        int right = heights.length - 1;
        int res = 0;

        while (left < right) {
            // calculate the area by finding the minimum and multiplying by
            // distance between right and left
            // l = 1, h = 6 equals 1 * 6 - 1 = 5
            int area = Math.min(heights[left], heights[right]) * (right - left);
            // compare res and current area
            res = Math.max(res, area);
            // if left pointer is less than right pointer increment left
            if (heights[left] < heights[right]) {
                left++;
            // if left pointer is greater than right pointer decrement right
            } else {
                right--;
            }
        }
        // return result
        return res;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:

        stack = []
        max_area = 0

        for i, h in enumerate(heights):
            # if stack is empty or current height is greater than heigh in stack
            if not stack or heights[i] > heights[stack[-1]]:
                stack.append(i)
            else:
                # while stack is not empty and current height is less than height in stack
                while stack and heights[i] <= heights[stack[-1]]:
                    # get height from stack
                    height = heights[stack.pop()]
                    # if the stack is empty then widhth = current index
                    if not stack:
                        width = i
                    # else calculate width, -1 for inclusion only
                    else:
                        width = i - stack[-1] - 1
                    # get current area
                    area = width * height
                    # get max area
                    max_area = max(area, max_area)
            stack.append(i)

        while stack:
            height = heights[stack.pop()]
            if not stack:
                width = len(heights)
            else:
                width = len(heights) - stack[-1] - 1
            # get current area
            area = width * height
            # get max area
            max_area = max(area, max_area)

        return max_area